### ISSUE TITLE
Pipeline: Unable to create Enrichment Streams

### ISSUE THREAD

**Issue by gaby** [openobserve/openobserve#5446]
### Which OpenObserve functionalities are the source of the bug?

pipeline

### Is this a regression?

Yes

### Description

Steps:

- Add Source -> Logs Stream
- Add Query to Filter on Field
- Add Output Stream:
 -> [x] Create a new stream
 -> Enrichment Stream
 -> Set Name
 -> Click save
     -> Error will popup saying an Enrichment Stream cant be created.

Details were discussed on Slack here: https://zincobserve.slack.com/archives/C04QC53TPD2/p1734799194932969?thread_ts=1734799194.932969&cid=C04QC53TPD2

I was told that @taimingl confirmed it's a bug.

### Please provide a link to a minimal reproduction of the bug

N/a

### Please provide the version you discovered this bug in (check about page for version information)

```true
v0.14.0 using Docker on Linux
```

### Anything else?

N/a

**Comment by gaby** [comment]
@taimingl Was there any updates on this bug?

**Comment by taimingl** [comment]
Hi @gaby, I had been working on something else these past two weeks and just started working through various pipeline issues yesterday. I will try to work on this one this week.

**Comment by gaby** [comment]
@taimingl Sounds good. I also noticed it happens in the `Streams` page 

Streams -> Add Stream -> Select Enrichment Table, it throws an error

**Comment by gaby** [comment]
@prabhatsharma @hengfeiyang @taimingl Is there an ETA for this? I'm on `v0.14.2` and it stills happens since the option was introduced.

You can't create enrichment tables from:
- Pipelines
- Streams Page

**Comment by taimingl** [comment]
hey @gaby, sorry about the delay. I'll pick this one up today.

**Comment by gaby** [comment]
> hey [@gaby](https://github.com/gaby), sorry about the delay. I'll pick this one up today.

Awesome, thank you! 💪

**Comment by gaby** [comment]
@taimingl This was closed, but #5854 didnt get merged

**Comment by taimingl** [comment]
@gaby, I think @nikhilsaikethe wants to combine the mentioned PR along with other UI improvements that he's currently working on. I reopened the issue. @nikhilsaikethe Please mention this issue in your pr once it's ready.

**Comment by gaby** [comment]
@nikhilsaikethe Did the work for this got lost? I never saw another PR linked to the issue. 😂

**Comment by nikhilsaikethe** [comment]
No @gaby I am currently working on pipeline improvements will merge along with that , here is the issue list
https://github.com/openobserve/openobserve/issues/5886
I will mention  this issue as well [Pipeline: Unable to create Enrichment Streams](https://github.com/openobserve/openobserve/issues/5446#top)
#5446

**Comment by gaby** [comment]
@nikhilsaikethe @taimingl Was excited to try v0.14.3, but the  option was just removed and moved to scheduled pipelines.

My use-case is to create an Enrichment Table in my real-time pipeline.

When creating a scheduled pipeline you also can't use a source stream from another pipeline.

**Comment by nikhilsaikethe** [comment]
@gaby  let me discuss about this issue / use case with the team and come back to you

**Comment by prabhatsharma** [comment]
Creating enrichment table from real time pipelines is not a good use case. Log streams have billions of records incoming regularly. Enrichment tables are supposed to be small.

Scheduled pipelines are the way to go

**Comment by gaby** [comment]
@prabhatsharma That's not everyone's use-case. OpenObserve is trying to dictate weird use-cases.

The option was there, but broken. Then it was removed and moved to Schedule Pipelines.

**Comment by gaby** [comment]
As a user I want to be able to use options anywhere. But the User experience is different between pages on OpenObserve.

Can use Feature A in page 1-3, but not 4-5
Dropdown has options 1-3 in page 1, but option 1-2 in page 3.

This happens with stream types and filtering options.

**Comment by prabhatsharma** [comment]
Not trying to dictate anything. 

Help me understand your use case. Convince me to build. So far I am not convinced.

**Comment by gaby** [comment]
Use-Case:

- We have several tasks that generate metadata about a system. This metadata is shipped as results along with other tools.
- Stream gets metadata from tasks. Some of this data we want to convert into an Enrichment table. Data arrives in random intervals, thus why the real-time use case.
  - We currently have a realtime pipeline for this Stream to drop some fields and process other fields.
- Enrichment table will be used by another stream to enrich it's data (ex: mapping ipv4 addresses to hostnames).

We tried adding that data into a stream and then running a query during ingest, but it's not efficient. The data is mostly mappings of values:
- ipv4 -> hostname
- location -> node
- uid -> username

The data is not static, thus the need for an Enrichment Table.

**Comment by gaby** [comment]
@prabhatsharma @nikhilsaikethe I was able to solve my use-case using `Schedule Pipelines` but ran into a bunch of bugs with that feature and pipelines.

- Using a `Query` node:
  - select a stream
  - click on a field, and it just gets appended to the query at the end
  - clicking run query throws a message saying to add a filter
- After saving a `Query` node:
  - click edit, and it doesn't populate the "stream" dropdown. The query stops.
 - When setting output into an `Enrichment Stream`, it forces a `_timestamp` field.
   - This causes duplicated entries in the enrichment table
   - Adding a function before output to drop `_timestamp` doesn't do anything. It gets auto added once data is output to the enrichment table.
  - The `Schedule Pipeline` table should say the last time a pipeline ran. It should also have a `play` button to trigger the pipeline.
  
Now that I have an enrichment table, I have no idea how to use it 😂

**Comment by nikhilsaikethe** [comment]
@gaby 
```
select a stream
click on a field, and it just gets appended to the query at the end
clicking run query throws a message saying to add a filter 
```
1. this if you write a query and select any stream this error should not come if this is coming can you please share me the screenshot so that it can be useful to debug
2. FYI when you click on + icon any field it gets added to the query --> here 2 scenarios one is if you place your cursor at particular place in the query it gets added there with a space at (prefix and suffix) , second is if not then it will get appended to the end of the query


`click edit, and it doesn't populate the "stream" dropdown. The query stops.
`
1. this autopopulate stream was removed intentionally but I will discuss with the team and let you know about this 


`The Schedule Pipeline table should say the last time a pipeline ran. It should also have a play button to trigger the pipeline.
`
1. the above feature we will work on it 

cc: @taimingl

**Comment by gaby** [comment]
@nikhilsaikethe You are right. If I write a query it doesn't happen, but it's not user friendly.

Everywhere else on O2, you select a stream and add fields. In this UI its completely different.

Same with the stream not getting auto-selected after clicking "Edit".

I would argue that this UI should only have the SQL query field and removing the "stream" and "fields" dropdown on the left. It's very confusing.

Thoughts on the `_timestamp` field being added to enrichment streams? I only have 150 values, and ended with a table with +1000 cause of this.

**Comment by nikhilsaikethe** [comment]
@gaby 
couple of fixes incoming 
`Same with the stream not getting auto-selected after clicking "Edit".
`
1. from now when you edit the stream,  will be auto selected 


` Everywhere else on O2, you select a stream and add fields. In this UI its completely different.
`
2. the initial query is auto filled right once you select a stream its in sync with the logs page only but if I understood you wrong please correct me .


`I would argue that this UI should only have the SQL query field and removing the "stream" and "fields" dropdown on the left. It's very confusing.`
3. why we have introduced is that while writing sql query user should be able to know about what are the fields that are there in the selected stream and also able to check the query by clicking run query so that they are getting the expected result for the query. 

`Thoughts on the _timestamp field being added to enrichment streams? I only have 150 values, and ended with a table with +1000 cause of this.`
regarding this let me comeback to you after discussing

**Comment by gaby** [comment]
Related https://github.com/openobserve/openobserve/pull/6188

**Comment by nikhilsaikethe** [comment]
Yes you are right

**Comment by gaby** [comment]
@nikhilsaikethe Thanks! 💪 Only thing left is the `_timestamp`. It seems like it shouldn't be there, specially if the table is only used for enrichment.

**Comment by gaby** [comment]
@nikhilsaikethe Reading the docs, it doesnt seem like enrichment tables are suppose to have a `_timestamp`.

See here https://openobserve.ai/docs/user-guide/functions/enrichment/

**Comment by nikhilsaikethe** [comment]
@gaby so when you select any enrichment_tables , you dont want to see _timestamp filed , am I correct??

**Comment by gaby** [comment]
@nikhilsaikethe Correct. If you upload a csv file it doesn't have a _timestamp field.

If you create an enrichment table via pipelines a _timestamp field is added. This causes entries with the same data to be duplicated.

Ex:

_timestamp, name, lastname
- 1234567890, john, smith
- 1234555566, john, smith
- 1234325325, john, doe

I end up with 2 entries for John Smith in the Enrichment table because the _timestamp field is being added.

**Comment by nikhilsaikethe** [comment]
@gaby let me reproduce that , give me some time

**Comment by nikhilsaikethe** [comment]
@gaby one more thing to reproduce that I am using a query node , a function that will drop the _timestamp field , and the destination node with enrichment_tables

**Comment by gaby** [comment]
> [@gaby](https://github.com/gaby) one more thing to reproduce that I am using a query node , a function that will drop the _timestamp field , and the destination node with enrichment_tables

That sounds correct. Then go to "Pipelines -> Enrichment Tables", then click the 🔍 icon next to your new enrichment table. You will see way more records than you expected and the _timestamp field.

**Comment by nikhilsaikethe** [comment]
@gaby what is the frequency you have set for the query node??

**Comment by gaby** [comment]
@nikhilsaikethe 60 minutes. Data is ingested every 45-60 minutes. But it changes like once or twice a day.

**Comment by nikhilsaikethe** [comment]
got it , Thank you for the information

**Comment by gaby** [comment]
I believe this is working fine, as of right now

**Comment by gaby** [comment]
@nikhilsaikethe I was running integrity checks, and realized that because its adding a `@timestamp` field I have multiple entries of the same Key with different values in my Enrichment Table.

When I query the table from VRL it returns the first Key it finds, which could have different values.

So, reopening this since that is a problem

**Comment by hengfeiyang** [comment]
@gaby sorry, i didn't get the issue, EnrichmentTable also is a stream, it always have _timestamp column, even you upload a csv file, it always have a _timestamp. why _timestamp is a problem?

**Comment by gaby** [comment]
@hengfeiyang There is a difference between uploading `once` a CSV enrichment table, and having one that gets updates constantly.

CSV:
- You will have 1 entry per `key/value` with a generated `_timestamp`

Enrichment Stream:
- You can have the same `key/value` pair with multiple `_timestamps` as values are not replaced when new data comes in. When the `value` changes, you will end up with:

```
1234567890, foo, bar
1234567899, foo, baz
```

How does VRL know which value to return now for key `foo`?

**Comment by hengfeiyang** [comment]
it is not about timestamp, because your csv file doesn’t have duplicate records, it is not about we remove the duplicate, we also add timestamp when you upload csv, you can try upload a csv with duplicate records, at the end, you also will have multiple records of same value. Or you used append=false mod, it will delete the old data when you upload again, @taimingl do we have append=false mod for pipeline for enrichment?

**Comment by gaby** [comment]
Yes, pipeline has an option for append True/False. I'm don't remember what I have mine set. Will check and report back tomorrow.

**Comment by gaby** [comment]
Disabling "Append" seems to have fixed the issue.
